src/test/java/fi/helsinki/cs/tmc/core/commands/DownloadAdaptiveExerciseTest.java

```diff
@@ -0,0 +1,144 @@
+        langs = spy(new TaskExecutorImpl());
+        TmcSettingsHolder.set(settings);
+        TmcLangsHolder.set(langs);
+        arithFuncsTempDir = testFolder.getRoot().toPath().resolve("arith_funcs");
+        command = new DownloadAdaptiveExercise(mockObserver, factory);
+
+        doCallRealMethod().when(langs).extractProject(any(Path.class), any(Path.class));
+        mockExerciseOne.setName("ex1");
+        mockExerciseOne.setCourseName("course1");
+
+    }
+
+    @Test
+    public void checkExerciseZipUrl() throws Exception {
+        setUpMocks();
+        Exercise exercise = command.call();
+    }
+
+    @Test
+    public void testDownloadAndExtractSuccess() throws Exception {
+        setUpMocks();
+
+        Exercise exercise = command.call();
+
+        verify(factory).getAdaptiveExercise();
+        verify(factory).getDownloadingExerciseZipTask(mockExerciseOne);
+
+        verifyNoMoreInteractions(factory);
+
+        assertTrue(Files.exists(arithFuncsTempDir));
+    }
+
+    @Test
+    public void testDownloadAndExtractFailure() throws Exception {
+        setUpMocks();
+        when(mockGetAdaptiveExercise.call()).thenReturn(null);
+
+        Exercise exercise = command.call();
+
+        verify(factory).getAdaptiveExercise();
+
+        verifyNoMoreInteractions(factory);
+
+    }
+
+    private void setUpMocks() throws Exception {
+        verifyZeroInteractions(langs);
+
+        when(factory.getAdaptiveExercise()).thenReturn(mockGetAdaptiveExercise);
+        when(mockGetAdaptiveExercise.call()).thenReturn(mockExerciseOne);
+
+        when(mockExerciseOne.getExtractionTarget(any(Path.class))).thenReturn(arithFuncsTempDir);
+        when(settings.getTmcProjectDirectory()).thenReturn(testFolder.getRoot().toPath());
+
+        when(factory.getDownloadingExerciseZipTask(mockExerciseOne))
+                .thenReturn(
+                        new Callable<byte[]>() {
+                            @Override
+                            public byte[] call() throws Exception {
+                                return Files.readAllBytes(
+                                        TestUtils.getZip(this.getClass(), "arith_funcs.zip"));
+                            }
+                        });
+    }
+
+    @Test
+    public void testDownloadAndExtractSuccessWithRealZip() throws Exception {
+        verifyZeroInteractions(langs);
+        TmcServerCommunicationTaskFactory realFactory = new TmcServerCommunicationTaskFactory();
+        assertNotNull(TmcSettingsHolder.get());
+        command = new DownloadAdaptiveExercise(mockObserver, realFactory);
+        Path testPath = Paths.get(System.getProperty("user.dir"));
+
+        when(settings.getTmcProjectDirectory()).thenReturn(testPath);
+
+        Exercise exercise = command.call();
+        assertTrue(Files.exists(testPath.resolve(exercise.getCourseName())));
+        FileUtils.deleteDirectory(testPath.resolve(exercise.getCourseName()).toFile());
```

Comment:
Do not delete stuff from my home, you can't know if that directory already existed. Use `createTemporaryDirectory` (or whatever it's called, the version that creates the dir in sys.temp) from `Files` to get a directory that is safe to delete later.